[IsDeprecatedWeakRefSmartPointerException] Make SampleBufferDisplayLayerManager ref-counted

RupinMittal opened the pull request:
#### 7fd8860065303c1f5a14252916b24a1b5e57a27f
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make SampleBufferDisplayLayerManager ref-counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=281248">https://bugs.webkit.org/show_bug.cgi?id=281248</a>
<a href="https://rdar.apple.com/137706376">rdar://137706376</a>

Reviewed by Ryosuke Niwa.

SampleBufferDisplayLayerManager is owned by GPUProcessConnection, so we forward
the ref/deref&apos;ing to GPUProcessConnection.

* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::sampleBufferDisplayLayerManager):
(WebKit::GPUProcessConnection::protectedSampleBufferDisplayLayerManager):
(WebKit::GPUProcessConnection::dispatchMessage):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::~SampleBufferDisplayLayer):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayerManager.cpp:
(WebKit::SampleBufferDisplayLayerManager::SampleBufferDisplayLayerManager):
(WebKit::SampleBufferDisplayLayerManager::ref const):
(WebKit::SampleBufferDisplayLayerManager::deref const):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayerManager.h:

Canonical link: <a href="https://commits.webkit.org/285025@main">https://commits.webkit.org/285025@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6cecbbadb1d958f387b764fe5e0ef6b69fa40a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71210 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75317 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22415 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58423 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22234 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56287 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14757 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45992 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61368 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36722 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42665 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20755 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77039 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15444 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18375 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64005 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15486 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61403 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63981 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15773 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12129 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5758 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46423 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1202 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47494 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48777 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47236 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -188,10 +188,15 @@ void GPUProcessConnection::didReceiveInvalidMessage(IPC::Connection&, IPC::Messa
 SampleBufferDisplayLayerManager& GPUProcessConnection::sampleBufferDisplayLayerManager()
```
I added a protected version of this to fix the one and only call site that became unsafe as a result of `SampleBufferDisplayLayerManager` now being ref-counted. That means that this non-protected `sampleBufferDisplayLayerManager()` is called only in the protected version and no where else. Should we still keep this non-protected version the way I have here?

cdumez replied:
Yes, I would keep in case someone needs it in the future.

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make SampleBufferDisplayLayerManager ref-counted
https://bugs.webkit.org/show_bug.cgi?id=281248
rdar://137706376

Reviewed by Ryosuke Niwa.

SampleBufferDisplayLayerManager is owned by GPUProcessConnection, so we forward
the ref/deref'ing to GPUProcessConnection.

* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::sampleBufferDisplayLayerManager):
(WebKit::GPUProcessConnection::protectedSampleBufferDisplayLayerManager):
(WebKit::GPUProcessConnection::dispatchMessage):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::~SampleBufferDisplayLayer):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayerManager.cpp:
(WebKit::SampleBufferDisplayLayerManager::SampleBufferDisplayLayerManager):
(WebKit::SampleBufferDisplayLayerManager::ref const):
(WebKit::SampleBufferDisplayLayerManager::deref const):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayerManager.h:

Canonical link: https://commits.webkit.org/285025@main
